### ISSUE TITLE
Investigate Potential Thread Starvation

### ISSUE THREAD

**Issue by ariasdiniz** [ariasdiniz/macaw_framework#64]
**Description**
During load testing, the framework exhibited potential thread starvation when the rate limiting and caching middlewares were enabled and the response size was over 70kb. After disabling these two middlewares, the application's performance improved substantially, indicating that one (or both) of these components may be holding onto threads longer than necessary during the handling of large responses.

**Steps to Reproduce**
1. Enable rate limiting and caching middlewares.
2. Run a Postman runner with 10,000 iterations, with each iteration involving a write, a patch, and two `findAll` operations in the database.
3. The find all requests should generate a response over 70kb.
4. Observe thread utilization and response times.

**Expected behavior**
The application should be able to handle a high volume of requests and large responses without experiencing significant latency or thread starvation.

**Actual behavior**
The application experienced potential thread starvation, and latency increased to minutes when the rate limiting and caching middlewares were enabled and the response size was over 70kb.

**Additional context**
- The application is hosted on a free EC2 instance provided by AWS.
- The load test involved significant database operations, including writes and fetches.

We need to investigate how these two middlewares are handling their threads, particularly when processing large responses, and ensure they're releasing them back to the pool in a timely manner. Also, we need to make sure these middlewares are thread-safe and capable of handling high concurrency loads without negatively impacting performance.

**Comment by ariasdiniz** [comment]
**Update:**

During our investigation into the performance issue we've discovered that the root cause of the problem is related to thread management. Specifically, our threads seem to be dying prematurely when the rate limiting and caching middlewares are enabled and the response size is over 70kb.

Given the fact that our application's performance improved substantially when these middlewares were disabled, we have determined that the issue is not due to inherent limitations of our server resources, but rather to how these two components are managing their threads.

We are currently in the process of conducting a deeper analysis of these middlewares to understand why they are not releasing their threads back to the pool in a timely manner. Our goal is to ensure that these middlewares are able to handle large responses effectively, without any unnecessary latency or thread starvation.

In addition, we are examining the thread-safety of these middlewares to ensure that they are capable of handling high concurrency loads without negatively impacting performance. This includes investigating any potential synchronization issues that might be causing the threads to die prematurely.

**Comment by ariasdiniz** [comment]
Issue Status: Partially Resolved

Update:

As we delved further into the performance issue, we discovered an interesting correlation between thread starvation and the size of request payloads. We found that the issue was more visible for requests with larger payloads due to their inherently longer processing time. This longer processing time provided a wider window for forced connection closes on the client side, leading to premature thread termination.

To address this issue, we have implemented a fix that optimizes thread management, specifically for large payloads. The fix ensures that even in cases of forced client-side connection termination, our server-side threads are properly managed and released back into the pool in a timely manner, thereby preventing starvation.

However, we want to stress that while the fix has shown promising results in our initial tests, we will continue to monitor its performance closely in a production environment to ensure its effectiveness. We also recognize that there might be other factors contributing to thread starvation, and will thus keep the issue open as we continue our investigations.

We appreciate your understanding and patience as we strive to resolve this issue fully.
